compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/dnd/ExternalDrag.desktop.kt

```diff
@@ -0,0 +1,459 @@
+            val dropTarget = window.dropTarget as? AwtWindowDropTarget ?: return@onDispose
+            val handleIdToRemove = componentDragHandleId ?: return@onDispose
+            dropTarget.stopDragHandling(handleIdToRemove)
+        }
+    }
+
+    Modifier
+        .onGloballyPositioned { position ->
+            // provide new component bounds to Swing to properly detect drag events
+            val dropTarget = window.dropTarget as? AwtWindowDropTarget
+                ?: return@onGloballyPositioned
+            val handleIdToUpdate = componentDragHandleId ?: return@onGloballyPositioned
+            val componentBounds = position.boundsInWindow()
+            dropTarget.updateComponentBounds(handleIdToUpdate, componentBounds)
+        }
+}
+
+/**
+ * Provides a way to subscribe on external drag for given [window] using [installComponentDragHandler]
+ *
+ * [Window] allows having only one [DropTarget], so this is the main [DropTarget] that handles all the drag subscriptions
+ *
+ * @VisibleForTesting
+ */
+@OptIn(ExperimentalComposeUiApi::class)
+internal class AwtWindowDropTarget(
+    private val window: Window
+) : DropTarget(window, DnDConstants.ACTION_MOVE, null, true) {
+    private var idsCounter = 0
+
+    // all components that are subscribed to external drag and drop for the window
+    private val handlers = mutableMapOf<Int, ComponentDragHandler>()
+
+    // bounds of all components that are subscribed to external drag and drop for the window
+    private val componentBoundsHolder = mutableMapOf<Int, Rect>()
+
+    // drag coordinates used to detect that drag entered/exited components
+    private var windowDragCoordinates: Offset? = null
+
+    // @VisibleForTesting
+    val dragTargetListener = AwtWindowDragTargetListener(
+        window,
+        // notify components on window border that drag is started.
+        onDragEnterWindow = { newWindowDragCoordinates ->
+            forAllPositionedComponents { handler, componentBounds ->
+                handleDragEvent(
+                    handler,
+                    oldComponentBounds = componentBounds, currentComponentBounds = componentBounds,
+                    oldDragCoordinates = null, currentDragCoordinates = newWindowDragCoordinates
+                )
+            }
+            windowDragCoordinates = newWindowDragCoordinates
+        },
+        // drag moved inside window, we should calculate whether drag entered/exited components or just moved inside them
+        onDragInsideWindow = { newWindowDragCoordinates ->
+            val oldDragCoordinates = windowDragCoordinates
+            windowDragCoordinates = newWindowDragCoordinates
+            forAllPositionedComponents { handler, componentBounds ->
+                handleDragEvent(
+                    handler,
+                    oldComponentBounds = componentBounds, currentComponentBounds = componentBounds,
+                    oldDragCoordinates, newWindowDragCoordinates
```

Comment:
Not sure about the style-guide, but I personally prefer to either put all arguments on one line, or one argument per line. @igordmn can say definitively whether it's important.

Reply:
Hmm, the [Kotlin style-guide](https://kotlinlang.org/docs/coding-conventions.html#method-calls) says to put related arguments on the same line, so maybe I'm wrong here.

Reply:
Yes, I did accroding to kotlin style-guide